viewing_party/party.py

```diff
@@ -1,23 +1,166 @@
+        if value > max_value:
+            max_value = value
+            max_key = key
+    return max_key
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    list_watched = []
+    for movie in user_data["watched"]:
+        list_watched.append(movie)
+
+    friends_watched_set = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_set.add(movie["title"])
+
+    list_unique = []
+    for movie in list_watched:
+        if movie["title"] not in friends_watched_set:
+            list_unique.append(movie)
+    return list_unique
+
+
+def get_friends_unique_watched(user_data):
+    set_watched = set()
+    for movie in user_data["watched"]:
+        set_watched.add(movie["title"])
+
+    friends_watched_dict = {}
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_dict[movie["title"]] = movie
+
+    list_unique = []
+    for key, value in friends_watched_dict.items():
+        if key not in set_watched:
+            list_unique.append(value)
+
+    return list_unique
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    set_watched = set()
+    for movie in user_data["watched"]:
+        set_watched.add(movie["title"])
+
+    friends_watched_list = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_list.append(movie)
+
+    list_only_friends_watched = []
+    for movie in friends_watched_list:
+        if movie["title"] not in set_watched:
+            list_only_friends_watched.append(movie)
+
+    recs_list = []
+    for movie in list_only_friends_watched:
+        if movie["host"] in user_data["subscriptions"]:
+            recs_list.append(movie)
+
+    return recs_list
+
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    most_watched_genre = get_most_watched_genre(user_data)
+
+    set_watched = set()
+    for movie in user_data["watched"]:
+        set_watched.add(movie["title"])
+
+    friends_watched_list = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_list.append(movie)
+
+    list_only_friends_watched = []
+    for movie in friends_watched_list:
+        if movie["title"] not in set_watched:
+            list_only_friends_watched.append(movie)
+
+    rec_by_genre_list = []
+    for movie in list_only_friends_watched:
+        if movie["genre"] == most_watched_genre:
+            rec_by_genre_list.append(movie)
+    return rec_by_genre_list
+
+
+def get_rec_from_favorites(user_data):

```

Comment:
Nice implementation, we could've also used `get_unique_watched` in both wave 5 functions as a helper function